pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,578 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'name' cannot be empty"
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'namespace' cannot be empty"
+	}
+
+	return builder
+}
+
+// WithAllAddonsDisabled disables all addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithAllAddonsDisabled() *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	enablement := false
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the all addons disabled", builder.Definition.Name)
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Grafana == nil {
+		builder.Definition.Spec.Addons.Grafana = &istioV2.GrafanaAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Kiali == nil {
+		builder.Definition.Spec.Addons.Kiali = &istioV2.KialiAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Jaeger == nil {
+		builder.Definition.Spec.Addons.Jaeger = &istioV2.JaegerAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Prometheus == nil {
+		builder.Definition.Spec.Addons.Prometheus = &istioV2.PrometheusAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Stackdriver == nil {
+		builder.Definition.Spec.Addons.Stackdriver = &istioV2.StackdriverAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.ThreeScale == nil {
+		builder.Definition.Spec.Addons.ThreeScale = &istioV2.ThreeScaleAddonConfig{}
+	}
+
+	disableAllAddons := &istioV2.AddonsConfig{
+		Prometheus: &istioV2.PrometheusAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+		Jaeger: &istioV2.JaegerAddonConfig{
+			Install: &istioV2.JaegerInstallConfig{
+				Ingress: &istioV2.JaegerIngressConfig{
+					Enablement: istioV2.Enablement{
+						Enabled: &enablement,
+					},
+				},
+			},
+		},
+		Grafana: &istioV2.GrafanaAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+		Kiali: &istioV2.KialiAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+	}
+
+	builder.Definition.Spec.Addons = disableAllAddons
+
+	return builder
+}
+
+// WithGrafanaAddon adds grafana addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithGrafanaAddon(
+	enablement bool,
+	install *istioV2.GrafanaInstallConfig,
+	address *string) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Grafana addons defined: enablement %s, install %s, "+
+			"address %s", builder.Definition.Name, enablement, install, address)
+
```

Comment:
Please check that install is not nil

	if install == nil {
		glog.V(100).Infof("The GrafanaInstallConfig of ServiceMeshControlPlane can not be empty")

		builder.errorMsg = "'jaegerAddonConfig' can not be empty for ServiceMeshControlPlane struct"
	}

	if builder.errorMsg != "" {
		return builder
	}

Reply:
it could be nil:
`apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: {{ istio_namespace }}
spec:
  addons:
    grafana:
      enabled: false
    jaeger:
      install:
        ingress:
          enabled: false
    kiali:
      enabled: false
    prometheus:
      enabled: false
  gateways:
    additionalIngress:
      amfmme1-istio-ingressgateway:
        enabled: true
        namespace: amfmme1
        runtime:
          container:
            resources:
              requests:
                cpu: "1"
                memory: 128Mi
          deployment:
            autoScaling:
              enabled: false
            replicas: 3
        service:
          metadata:
            labels:
              amfmme1: ingressgateway...`

Reply:
what do you mean??? For each one of the methods there were in params verification added?
in case someone will use With... method with no providing param it will fail on compilation no?
And, also, I do not see we do so for any With method